### PR TITLE
Ga4gh implementation with EmbeddedTokens

### PR DESCRIPTION
**Pull request by martin-kuba**

### CODE REVIEW

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/PerunAccessTokenEnhancer.java [hunk 1]
```diff
@@ -99,8 +99,8 @@ public OAuth2AccessToken enhance(OAuth2AccessToken accessToken, OAuth2Authentica
         JWTClaimsSet claims = builder.build();
 
         JWSAlgorithm signingAlg = jwtService.getDefaultSigningAlgorithm();
-        URI jku = URI.create(configBean.getIssuer() + JWKSetPublishingEndpoint.URL);
-        JWSHeader header = new JWSHeader(signingAlg, null, null, null, jku, null, null, null, null, null,
+//        URI jku = URI.create(configBean.getIssuer() + JWKSetPublishingEndpoint.URL);
```
Comment (pajavyskocil) [inline oidc-idp/src/main/java/cz/muni/ics/oidc/server/PerunAccessTokenEnhancer.java]: Delete the comment if it is not necessary

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+		long endx = System.currentTimeMillis();
+		System.out.println("signature verification time: " + (endx - startx));
+
+	}
+
+	private static Map<URL, RemoteJWKSet<SecurityContext>> remoteJwkSets = new HashMap<>();
+
+	private static boolean verifyJWT(String jwtString) throws ParseException, MalformedURLException, JOSEException {
+//		System.out.println();
+//		System.out.println("verifying a JWT ...");
```
Comment (pajavyskocil) [inline oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: Remove the comments

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		return signedJWT.verify(verifier);
+	}
+
+	private static String isoDateTime(long linuxTime) {
+		return DateTimeFormatter.ISO_LOCAL_DATE_TIME.format(ZonedDateTime.ofInstant(Instant.ofEpochSecond(linuxTime), ZoneId.systemDefault()));
+	}
+	private static void prettyPrintPayload(String payload) {
+		try {
+			JsonNode doc = jsonMapper.readValue(payload, JsonNode.class);
+//			System.out.println(jsonMapper.writerWithDefaultPrettyPrinter().writeValueAsString(doc));
```
Comment (pajavyskocil) [inline oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: Remove or uncomment the line

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+		ObjectMapper jsonMapper = new ObjectMapper();
+		JsonNode doc = jsonMapper.readValue(new File(userinfo), JsonNode.class);
+		JsonNode ga4gh = doc.get("ga4gh");
+		long startx = System.currentTimeMillis();
+		processRIClaim(ga4gh, "AcceptedTermsAndPolicies");
+		processRIClaim(ga4gh, "ResearcherStatus");
+		processRIClaim(ga4gh, "ControlledAccessGrants");
+		processRIClaim(ga4gh, "AffiliationAndRole");
+		long endx = System.currentTimeMillis();
+		System.out.println("signature verification time: " + (endx - startx));
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: This class is just a CLI utility for debugging, so it is.

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+	}
+
+	private static void processRIClaim(JsonNode ga4gh, String claimName) throws ParseException, JOSEException, MalformedURLException {
+		System.out.println();
+		System.out.println("---- " + claimName + " --------");
+		for (JsonNode jwtString : ga4gh.get(claimName)) {
+			String s = jwtString.asText();
+			if(!verifyJWT(s)) {
+				System.out.println("chyba v "+s);
```
[the remark this answers is not in the document]
Comment (martin-kuba) [reply oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: fixed

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/PerunAccessTokenEnhancer.java (hunk 1 above)
Comment (martin-kuba) [reply oidc-idp/src/main/java/cz/muni/ics/oidc/server/PerunAccessTokenEnhancer.java]: removed

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java (hunk 2 above)
Comment (martin-kuba) [reply oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: removed

File: oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (pajavyskocil) [reply oidc-idp/src/main/java/cz/muni/ics/oidc/server/elixir/GA4GHTokenParser.java]: OK

**Pull request merged**